### PR TITLE
Verify that program type is supported by attach provider

### PR DESCRIPTION
**Pull request by Alan-Jowett**
Attach provider gives the supported program type.
Link verifies that program type matches supported program type.

Resolves: #239

Signed-off-by: Alan Jowett <alanjo@microsoft.com>

### CODE REVIEW

File: libs/execution_context/ebpf_link.c [hunk 1]
```diff
@@ -97,6 +114,16 @@ ebpf_link_attach_program(ebpf_link_t* link, ebpf_program_t* program)
     link->program = program;
     ebpf_object_acquire_reference((ebpf_object_t*)program);
 
+    return_value = ebpf_program_get_properties(program, &program_parameters);
+    if (return_value != EBPF_SUCCESS) {
+        goto Done;
+    }
+
+    if (memcmp(&program_parameters.program_type, &link->program_type, sizeof(link->program_type)) != 0) {
```
Comment (shankarseal) [inline libs/execution_context/ebpf_link.c:122]: Minor: Since these are GUIDs, may be just a `==` is sufficient?

Comment (Alan-Jowett) [reply libs/execution_context/ebpf_link.c:122]: Nope:
https://docs.microsoft.com/en-us/cpp/error-messages/compiler-errors-1/compiler-error-c2088

C doesn't support == for structs.

File: tests/ext/drv/test_ext.c [hunk 2]
```diff
@@ -191,6 +193,13 @@ NTSTATUS
 _test_ebpf_extension_hook_provider_detach_client(_In_ void* provider_binding_context);
 
 // Test eBPF extension Hook NPI provider characteristics
+ebpf_attach_provider_data_t _test_ebpf_extension_attach_provider_data = {
+    {0xf788ef4a, 0x207d, 0x4dc3, {0x85, 0xcf, 0x0f, 0x2e, 0xa1, 0x07, 0x21, 0x3c}}};
```
Comment (shankarseal) [inline tests/ext/drv/test_ext.c:197]: this becomes a dupe of EBPF_PROGRAM_TYPE_TEST then?
If you are getting compiler errors referring to EBPF_PROGRAM_TYPE_TEST,
Maybe make `ebpf_program_type_t* supported_program_type;` a pointer and then initialize _test_ebpf_extension_hook_provider_data with  EBPF_PROGRAM_TYPE_TEST

File: netebpfext/ebpf_ext_attach_provider.h [hunk 3]
```diff
@@ -46,6 +47,7 @@ _Releases_lock_(registration) void ebpf_ext_attach_leave_rundown(
 ebpf_result_t
 ebpf_ext_attach_register_provider(
     _In_ const ebpf_attach_type_t* attach_type,
+    _In_ const ebpf_program_type_t* program_type,
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:50]: ebpf_program_load puts program_type before attach_type.  We should use consistent order of arguments.

File: netebpfext/ebpf_ext_attach_provider.h [hunk 4]
```diff
@@ -37,6 +37,7 @@ _Releases_lock_(registration) void ebpf_ext_attach_leave_rundown(
  * @brief Register as a attach type provider.
  *
  * @param[in] attach_type Attach type to register for.
+ * @param[in] program_type Permitted program type.
```
Comment (dthaler) [inline netebpfext/ebpf_ext_attach_provider.h:40]: update order to match order of args

**Review by dthaler — APPROVED** [review]

**Review by shankarseal — APPROVED** [review]

**Pull request merged**